# 第三次课：Python 运算符（续）

本节在上次课的基础上，系统梳理算术运算符、比较运算符、赋值运算符、逻辑运算符、位运算符、成员运算符与身份运算符。

## 1. 算术运算符复习

字符串与整数相乘（序列重复）、取模、取整除：

In [1]:
# 字符串与整数相乘的示例
x = "a"
y = 3
print("x*y =", x*y)
# 取模 % 的示例
x = 34
y = 5
print("x%y =", x%y)
# 取整除的示例
x = 12
y = 5
print("正数x//y =", x//y)
x = -12
y = 5
print("负数x//y =", x//y)


x*y = aaa
x%y = 4
正数x//y = 2
负数x//y = -3


字符串重复与拼接的组合运算：

In [2]:
x = "abc"
y = 4
print(x*y+x)


abcabcabcabcabc


`x*y` 得到 4 个 `"abc"`，再用 `+` 拼接一个 `"abc"`，共 5 个，即 `abcabcabcabcabc`。

多种算术运算符的混合运算：

In [3]:
x = 1
y = 2
print(x**y + x//y + x/y + x%y)


2.5


**解释：**

- `**` 幂 —— 返回 x 的 y 次幂，`x**y` = 1 的 2 次方 = 1
- `//` 取整除 —— 向下取接近商的整数，`x//y` = 1 除 2 取整，为 0
- `/` 除 —— x 除以 y，`x/y` = 0.5
- `%` 取模 —— 返回除法的余数，`x%y` = 1

合计：1 + 0 + 0.5 + 1 = **2.5**

## 2. 比较运算符

### `==`（等于）

In [4]:
# 比较是否相等 == 的示例
P_采购价格 = 205
s_标准价格 = 200
if (P_采购价格 == s_标准价格):
    print("采购价格 等于 标准价格")
else:
    print("采购价格 不等于 标准价格")


采购价格 不等于 标准价格


### `!=`（不等于）

`!=` —— 不等于，比较两个对象是否不相等。

In [5]:
# 比较是否不相等 != 的示例
P_采购价格 = 205
s_标准价格 = 200
if (P_采购价格 != s_标准价格):
    print("采购价格 等于 标准价格")
else:
    print("采购价格 不等于 标准价格")


采购价格 等于 标准价格


> ⚠️ **注（原文如此）：** 205 != 200 条件成立，进入 `if` 分支，但该分支的打印文本写的是"等于"，与条件逻辑相反。这里按讲义原样保留；正确写法应将两个分支的文本互换。

### 比较运算的结果可以参与算术运算

比较运算的结果是布尔值 `True`（数值 1）/ `False`（数值 0），可直接参与加减：

In [6]:
x = 32
y = 35
print((x==y) + 6 + (x!=y) + (x>y) + (x<=y))


8


**解释：**

| 表达式 | 判断 | 结果 | 数值 |
|---|---|---|---|
| `(x==y)` | 32 等于 35 吗？ | False | 0 |
| `6` | —— | —— | 6 |
| `(x!=y)` | 32 不等于 35 吗？ | True | 1 |
| `(x>y)` | 32 大于 35 吗？ | False | 0 |
| `(x<=y)` | 32 小于或等于 35 吗？ | True | 1 |

合计：0 + 6 + 1 + 0 + 1 = **8**

## 3. 赋值运算符

### `*=`（乘法赋值运算符）

`c *= a` 等效于 `c = c * a`。

In [7]:
# 赋值运算符 "*=" 示例
S_标准价格 = 200
F_价格系数 = 1.2
S_标准价格 *= F_价格系数
print(S_标准价格)


240.0


**解释：** `S_标准价格 *= F_价格系数`，即 `S_标准价格 = S_标准价格 * F_价格系数` = 200 × 1.2 = 240（浮点数显示为 240.0）。

### `-=`（减法赋值运算符）

`c -= a` 等效于 `c = c - a`。

In [8]:
# 赋值运算符 "-=" 示例
S_标准成本 = 309
H_人工成本 = 12
S_标准成本 -= H_人工成本
print(S_标准成本)


297


**解释：** `S_标准成本 -= H_人工成本`，即 `S_标准成本 = S_标准成本 - H_人工成本` = 309 − 12 = 297。

### `:=`（海象运算符，Python 3.8+）

在表达式内部完成赋值，避免重复书写：

In [9]:
# 传统的代码
my_list = [1, 2, 3, 4, 5]
count = len(my_list)
if count > 3:
    print("传统代码示例 - 错误！my_list 中 元素太多")

# 当转换为海象运算符时...
# 注意：需要 Python 3.8+ 版本
if (count := len(my_list)) > 3:
    print("海象运算符示例 - 错误！my_list 中 元素太多")


传统代码示例 - 错误！my_list 中 元素太多
海象运算符示例 - 错误！my_list 中 元素太多


## 4. 逻辑运算符

### `and`（复习：短路求值）

In [10]:
a = 1      # 真值（非0）
b = 2
print(bool(a), bool(b), (a and b))
# 因为 a=1 是真值，所以返回 b 的值 2

a = 0      # 假值（0）
b = 9
print(bool(a), bool(b), (a and b))
# 因为 a=0 是假值，短路逻辑直接返回 a 的值 0，不计算 b


True True 2
False True 0


### `or`

第一个值为 `True` → 直接返回第一个值；第一个值为 `False` → 继续检查并返回第二个值。

In [11]:
# 逻辑运算 or 示例
# a 是 True，返回 a
a = 1
b = 2
print(bool(a), bool(b), (a or b))

# a 不为 True，返回 b
a = 0
b = 9
print(bool(a), bool(b), (a or b))


True True 1
False True 9


**解释：**

- `a = 1`：在 Python 中，非零数字被视为 True；`b = 2` 也是 True。
- `a or b`：因为 a（1）是真值，根据规则，直接返回 a 的值，即 1。
- `a = 0`：0 被视为 False；`b = 9` 是非零数字，被视为 True。
- `a or b`：因为 a（0）是假值，Python 会继续检查 b，并返回 b 的值，即 9。

### `not`（逻辑取反）

In [12]:
# 逻辑运算 not 示例
# a 是 True，返回 False
a = 3
print(bool(a), (not a))
# a 是 False，返回 True
a = 0
print(bool(a), (not a))


True False
False True


**解释：**

- `a = 3`：非零数字被视为真，`bool(a)` 结果为 True；`not a` 对真值取反，结果为 False。输出：`True False`
- `a = 0`：0 被视为假，`bool(a)` 结果为 False；`not a` 对假值取反，结果为 True。输出：`False True`

## 5. 位运算符

按位与 `&`、按位或 `|`、左移 `<<` 的组合运算：

In [13]:
a = 3
b = 1
print(((a&b)|b)<<1)


2


**解释（将数字转换为二进制逐步拆解）：**

初始值：`a = 3`（二进制 `0011`），`b = 1`（二进制 `0001`）。运算顺序从内层括号向外层：

| 步骤 | 运算 | 二进制过程 | 结果 |
|---|---|---|---|
| 1 | 按位与 `a & b` 即 `3 & 1` | `0011 & 0001`：只有两个位都为 1 时结果才为 1 | 1 |
| 2 | 按位或 `1 \| 1` | 只要有一个位为 1，结果就为 1；(A & B) \| B 恒等于 B | 1 |
| 3 | 左移 `1 << 1` | 左移 1 位相当于乘以 2，即 1 × 2 = 2 | **2** |

所以最终输出为 **2**。

## 6. 成员运算符

`in`：判断元素是否在序列中；`not in`：判断元素是否不在序列中。

In [14]:
# in 示例
a = "营业收入"
b = ["营业收入", "利息收入", "已赚保费", "手续费及佣金收入"]
print("a in b =", a in b)
# not in 示例
c = "营业成本"
b = ["营业收入", "利息收入", "已赚保费", "手续费及佣金收入"]
print("c not in b =", c not in b)


a in b = True
c not in b = True


## 7. 身份运算符

`==` 比较值是否相等；`is` 比较内存地址（是否为同一个对象）：

In [15]:
# is 示例：int 类型时
x = 20
y = 20
print("x==y", x==y, "x is y =", x is y)
# is 示例：字符串类型时
x = "123"
y = "123"
print("x==y", x==y, "x is y =", x is y)
# is 示例：列表类型时
x = [1, 2, 3]
y = [1, 2, 3]
print("x==y", x==y, "x is y =", x is y)
# is 示例：元组类型时
x = (1, 2, 3)
y = (1, 2, 3)
z = x
print("x==y", x==y, "x is y =", x is y, "x is z =", x is z)


x==y True x is y = True
x==y True x is y = True
x==y True x is y = False
x==y True x is y = False x is z = True


**解释：**

- `==`：比较值是否相等；`is`：比较内存地址（是否为同一个对象）。
- **整数和小字符串**：Python 会进行驻留（interning），所以 `is` 返回 True。
- **列表**：每次创建都是新对象，内存地址不同，所以 `is` 返回 False。
- **元组**：虽然不可变，但分别创建时仍是不同对象，`x is y` 为 False；`z = x` 是引用同一对象，所以 `x is z` 为 True。

> 注：`is` 的结果依赖解释器的驻留优化，不同环境（如交互式与脚本模式）可能表现不同，建议在需要比较值时使用 `==`。